LLama.Unittest/KernelMemory/ITextTokenizerTests.cs

```diff
@@ -0,0 +1,81 @@
+using LLama.Common;
+using LLamaSharp.KernelMemory;
+using Microsoft.KernelMemory.AI;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Xunit.Abstractions;
+
+namespace LLama.Unittest.KernelMemory
+{
+    public abstract class ITextTokenizerTests
+    {
+        private readonly ITestOutputHelper _testOutputHelper;
+
+#pragma warning disable KMEXP00 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
+        protected ITextTokenizer? _generator;
+#pragma warning restore KMEXP00 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
+
+        protected InferenceParams _infParams;
+        protected LLamaSharpConfig _lsConfig;
+
+        public ITextTokenizerTests(ITestOutputHelper testOutputHelper)
+        {
+            _testOutputHelper = testOutputHelper;
+
+            _infParams = new() { AntiPrompts = ["\n\n"] };
+            _lsConfig = new(Constants.GenerativeModelPath) { DefaultInferenceParams = _infParams };
+
+            testOutputHelper.WriteLine($"Using model {Path.GetFileName(_lsConfig.ModelPath)}");
+        }
+
+        [Theory]
```

Comment:
Could you try adding this string https://github.com/SciSharp/LLamaSharp/blob/master/LLama.Unittest/StreamingTextDecoderTests.cs#L45 to your test cases, you'll probably get some of the weird behaviour with empty strings I mentioned above. At least then it will be documented by the tests.

Reply:
I get:
```
Using model llama-2-7b-chat.Q3_K_S.gguf
Tokens for '猫坐在垫子上 😀🤨🤐😏':
()( )()()(猫)()()(坐)(在)()()(垫)(子)(上)( )()()()(😀)()()()(🤨)()()()(🤐)()()()(😏)
```

Could this be a limitation of the generative model used for unit testing? llama-3 seems to do a bit better:
```
Using model Meta-Llama-3-8B-Instruct.Q8_0.gguf
Tokens for '猫坐在垫子上 😀🤨🤐😏':
()(猫)(坐在)()(垫)(子)(上)( 😀)()()(🤨)()()(🤐)()(😏)
```

I also noticed llama-2 has difficulties parsing the greek fragment and simply treats every letter as its own token, while llama-3 appears better at parsing it as a language:

```
Using model llama-2-7b-chat.Q3_K_S.gguf
Tokens for 'And a little bit of unicode για να κρατήσουμε τα πράγματα ενδιαφέροντα':
()( And)( a)( little)( bit)( of)( unicode)( )(γ)(ι)(α)( )(ν)(α)( )(κ)(ρ)(α)(τ)(ή)(σ)(ο)(υ)(μ)(ε)( )(τ)(α)( )(π)(ρ)(ά)(γ)(μ)(α)(τ)(α)( )(ε)(ν)(δ)(ι)(α)(φ)(έ)(ρ)(ο)(ν)(τ)(α)
```
vs
```
Using model Meta-Llama-3-8B-Instruct.Q8_0.gguf
Tokens for 'And a little bit of unicode για να κρατήσουμε τα πράγματα ενδιαφέροντα':
()(And)( a)( little)( bit)( of)( unicode)( για)( να)( κρα)(τή)(σ)(ουμε)( τα)( πρά)(γ)(ματα)( εν)(δια)(φέ)(ρον)(τα)
```

Reply:
The cause is that some individual characters are actually multiple tokens. The character `🤨` is the tokens `<0xF0><0x9F><0xA4><0xA8>` (using llama2 tokenizer, it'll vary between different tokenizers but the fundamental issue is the same). As an example if you call `GetTokens("🤨")` it will do:

1. Add `0xF0` to the StreamingTokenDecoder
2. Read() returns nothing, because that token can't be decoded on it's own
3. Add `0x9F` 
4. Read() returns nothing again
5. Add `0xA4`
6. Read() returns nothing again
7. Add `0xA8`
8. Read() finally returns "🤨"

The `GetTokens` method in KernelMemory assumes that each individual token has a string associated with it, but that's not true! That's why I commented it's bad API design by KernelMemory.

There are three ways I can see to handle this:
1. Remove the empty strings. This seems like the worst options to me, because then CountTokens will return a different count to what GetTokens returns!
2. Do as you've currently done, return empty strings for all the tokens that cannot be decoded. This way the count is correct, and the output is human readable.
3. Simply return the numeric value of every token, this way the output is less "weird", but completely unreadable to humans. In this case GetTokens would become:

```csharp
var tokens = context.Tokenize(text, special: true);
return tokens.Select(a => a.ToString()).ToArray();
```

I'm happy to go with 2 or 3, it's up to you.

Reply:
I'd say 2, if anything returning a List<string> of numbers seems very counter intuitive and as you say the KernelMemory implementation returns strings. Hopefully in the future things will become clearer,

I sneaked a call to GetTokens in their CountTokens unit test and here's what it currently does:

https://github.com/microsoft/kernel-memory/blob/main/extensions/LlamaSharp/LlamaSharp.FunctionalTests/LlamaSharpTextGeneratorTest.cs
```
warn: Microsoft.KernelMemory.AI.LlamaSharp.LlamaSharpTextGenerator[0]
      Tokenizer not specified, will use GPT4Tokenizer. The token count might be incorrect, causing unexpected errors

[...]
Tokens for '猫坐在垫子上 😀🤨🤐😏'
(猫)(猫)(猫)(坐)(坐)(在)(垫)(垫)(垫)(子)(上)( 😀)(🤨)(🤨)(🤨)(🤐)(🤐)(🤐)(😏)(😏)
```

Reply:
Huh, looks like they've gone for a slightly different approach! They're duplicating the tokens, so if tokens `[1, 2, 3]` decode to `A` they're outputing `["A", "A", "A"]` instead of `[" ", " ", "A"]`.

For now, I'm happy to merge this as is if you are? I imagine this is an area of Kernel Memory that's going to get a bit of a redesign in the future!